### PR TITLE
Return to batch norm dynamic batch size 1 and 2 kernels 

### PR DESCRIPTION
**Pull request by daniellowell**

### CODE REVIEW

File: src/batch_norm.cpp [hunk 1]
```diff
@@ -565,19 +570,52 @@ void bnBwdTrainSelectSingle(const Handle& handle,
...
+                handle.AddKernel(
+                    algo_name, network_config, program_name, kernel_name, vld, vgd, parms)(
+                    x, dy, dx, bnScale, dScale, dBias, epsilon, nhw, in_cstride, in_nstride);
+            }
         }
         else
         {
-            handle.AddKernel(algo_name, network_config, program_name, kernel_name, vld, vgd, parms)(
-                x, dy, dx, bnScale, dScale, dBias, epsilon, as_float(inhw));
+            if(useSaved)
```
Comment (muralinr) [inline src/batch_norm.cpp:607]: is it not possible to club "useSaved" and not saved options into one single call?

File: src/batch_norm.cpp [hunk 2]
```diff
@@ -565,19 +570,52 @@ void bnBwdTrainSelectSingle(const Handle& handle,
                             double epsilon,
                             ConstData_t savedMean,
                             ConstData_t savedInvVariance,
-                            float inhw)
```
Comment (muralinr) [inline src/batch_norm.cpp:568]: I see you are calculating inverse nhw in OCL code. is it possible to send it inhw directly?

File: src/include/miopen/batch_norm.hpp [hunk 3]
```diff
@@ -66,10 +67,14 @@ void bnBwdTrainSelectSingle(const Handle& handle,
                             double epsilon,
                             ConstData_t savedMean,
                             ConstData_t savedInvVariance,
-                            float inhw);
+                            float nhw,
```
Comment (muralinr) [inline src/include/miopen/batch_norm.hpp:70]: I see you are calculating inverse nhw in OCL code. is it possible to send it inhw directly?

File: test/bn_spatial_test.cpp [hunk 4]
```diff
@@ -1124,8 +1124,7 @@ struct batch_norm_spatial_driver : test_driver
         std::size_t n, c, h, w;
         std::tie(n, c, h, w) = miopen::tien<4>(input.desc.GetLengths());
 
-        if(n == 1 ||
-           ((h * w > 1024) && (input.desc.GetType() == miopenHalf) && (MIO_BN_USE_MIX_PREC == 0)) ||
+        if(((h * w > 1024) && (input.desc.GetType() == miopenHalf) && (MIO_BN_USE_MIX_PREC == 0)) ||
```
Comment (muralinr) [inline test/bn_spatial_test.cpp:1127]: Ok. got it.

File: test/bn_spatial_test.cpp [hunk 5]
```diff
@@ -1255,7 +1264,7 @@ struct batch_norm_spatial_driver : test_driver
 #if(MIO_BN_SP_TEST_DEBUG == 1)
         std::cout << "Running back propagation spatial recalc." << std::endl;
 #endif
-        this->tolerance = 80 * input.desc.GetElementSize();
+        this->tolerance = 80000 * input.desc.GetElementSize();
```
Comment (muralinr) [inline test/bn_spatial_test.cpp:1267]: What's reason for adjusting tolerance?  Is it due to smaller batch size?

File: test/network_data.hpp [hunk 6]
```diff
@@ -438,7 +438,45 @@ get_bn_spatial_inputs(int n = MIOPEN_TEST_DEFAULT_BATCH_SIZE_FACTOR)
         { pick_batch_size(32, n),  480,  128, 256 },
         { pick_batch_size(32, n),  528,  64,  128 },
         { pick_batch_size(770, n),  1,  8,  8 },
-        { pick_batch_size(770, n),  1024,  1,  1 }
+        { pick_batch_size(770, n),  1024,  1,  1 },
```
Comment (muralinr) [inline test/network_data.hpp:441]: Thanks for adding new test cases for smaller batch sizes.

**Review by muralinr — APPROVED** [review]

File: src/include/miopen/batch_norm.hpp (hunk 3 above)
Comment (daniellowell) [reply src/include/miopen/batch_norm.hpp:70]: In backwards we use both 1/nhw and nhw. I suppose I could send the both in. Maybe in another PR.

File: test/bn_spatial_test.cpp [hunk 7]
```diff
@@ -1172,9 +1171,9 @@ struct batch_norm_spatial_driver : test_driver
 #endif
         // this->tolerance = 80;
         // Debug values
-        // std::fill(input.begin(), input.end(), 1);
-        // std::fill(scale.begin(), scale.end(), 1);
-        // std::fill(shift.begin(), shift.end(), 1);
+        /* std::fill(input.begin(), input.end(), 1);
+        std::fill(scale.begin(), scale.end(), 1);
+        std::fill(shift.begin(), shift.end(), 1); */
```
Comment (atamazov) [inline test/bn_spatial_test.cpp:1176]: [Recommendation]
```suggestion
#if SELF_EXPLAINING_MACRO_NAME
        std::fill(input.begin(), input.end(), 1);
        std::fill(scale.begin(), scale.end(), 1);
        std::fill(shift.begin(), shift.end(), 1);
#endif
```
Remove `// Debug values` comment, and add somewhere near beginning of the source
```С++
#define SELF_EXPLAINING_MACRO_NAME 0
```

File: test/bn_3d_spatial_test.cpp [hunk 8]
```diff
@@ -1347,7 +1340,7 @@ struct batch_norm_3d_spatial_driver : test_driver
 #if(MIO_BN_SP_TEST_DEBUG == 1)
         std::cout << "Running back propagation spatial recalc." << std::endl;
 #endif
-        this->tolerance = 80 * input.desc.GetElementSize();
+        this->tolerance = 8000 * input.desc.GetElementSize();
```
Comment (atamazov) [inline test/bn_3d_spatial_test.cpp:1343]: [Question] Same question as https://github.com/ROCmSoftwarePlatform/MIOpen/pull/564/files#r517015425

File: test/bn_spatial_test.cpp (hunk 5 above)
Comment (atamazov) [reply test/bn_spatial_test.cpp:1267]: I am second to this question.

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. You can continue work as usual. But please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

**Comment by atamazov** [comment]
We need someone who will take ownership of this PR.

/cc @JehandadKhan @dagamayank

**Comment by muralinr** [comment]
[AMD Official Use Only - Internal Distribution Only]

Hi Artem,
I will own this PR.

Thanks,
Murali

**Comment by atamazov** [comment]
@muralinr Can you please elaborate the PR description (the topmost comment) a bit? What is the motivation, purpose of this PR, expected side effects etc, Bug fix or enhancement? This is important for reviewers.

**Comment by JehandadKhan** [comment]
This PR is stale and has not been updated in two years

**Comment by atamazov** [comment]
@muralinr Any objections? Or you would like to complete this?

**Comment by muralinr** [comment]
> @muralinr Any objections? Or you would like to complete this?

@atamazov  We are phasing out OCL batchnorm code soon.  No more legacy code changes.
